src/rsz/src/PreChecks.cc

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) 2023, Precision Innovations Inc.
+// All rights reserved.
+//
+// BSD 3-Clause License
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "PreChecks.hh"
+#include "rsz/Resizer.hh"
+#include "db_sta/dbNetwork.hh"
+#include "sta/Corner.hh"
+#include "sta/LibertyClass.hh"
+#include "utl/Logger.h"
+
+namespace rsz {
+
+using sta::LibertyCell;
+using sta::LibertyCellSeq;
+using utl::RSZ;
+
+PreChecks::PreChecks(Resizer* resizer) :
+        logger_(nullptr),
+        sta_(nullptr),
+        resizer_(resizer),
+        best_case_slew_computed_(false) {
+}
+
+void
+PreChecks::init()
+{
+  logger_ = resizer_->logger_;
+  sta_ = resizer_->sta_;
+}
+
+void PreChecks::checkSlewLimit(float ref_cap, float max_load_slew) {
+  // Ensure the max slew value specified is something the library can
+  // potentially handle
+  if (!best_case_slew_computed_ || ref_cap < best_case_slew_load_) {
+    LibertyCellSeq *equiv_cells = sta_->equivCells(resizer_->buffer_lowest_drive_);
+    float slew = resizer_->bufferSlew(resizer_->buffer_lowest_drive_, ref_cap, resizer_->tgt_slew_dcalc_ap_);
+    if (equiv_cells != nullptr) {
+      for (LibertyCell *buffer: *equiv_cells) {
+        slew = std::min(slew, resizer_->bufferSlew(buffer, ref_cap, resizer_->tgt_slew_dcalc_ap_));
+      }
+    }
+    best_case_slew_computed_ = true;
+    best_case_slew_load_ = ref_cap;
+    best_case_slew_ = slew;
+  }
```

Comment:
It seems like we can avoid best_case_slew_computed_, depending on how this code interacts with resizer.

Can we move this code to the init function, and just calculate it once? That would obviate the need for the best_case_slew_computed_ variable since we would just do it once during init.

If it's not possible to move it to init then I think we should wrap best_case_slew_load_ and best_case_slew_ in struct, and mark it as std::optional, and check for its presence. It would make it clearer to the future readers that this is really just a cache. I think it could potentially also make sense to move this into a private method.

```c++
private: 
struct LibrarySlewProperties {
  float best_case_slew_load_;
  float best_case_slew_load_;
};

LibrarySlewProperties getLibrarySlewProperties() {
  if (slew_properties_) {
    return *slew_properties_;
  }

  // Calculate and set slew_properties_
}

std::optional<LibrarySlewProperties> slew_properties_;
```

Reply:
@QuantamHD The problem here is that the best case slew depends on the load. I do not quite comprehend how we can compute it just once when load might need to depend on cell placement. It is a cache.... but a cache that is only valid as long as we don't get a lower best case load (when the cache becomes invalid). Too much abstraction does not necessarily lead to better code or more maintainable code ... but let me think about that some more.

Reply:
Sounds good, and generally agree on not adding unnecessary levels of abstraction.

I would generally like to see more structure, and separation of concerns in the codebase, but you're in the best position to judge the trade-off between simplicity and abstraction.

Reply:
@QuantamHD I looked into your suggestions and for now I think I will leave it as it is. We could theoretically do this check at a different level but that is most likely a larger project which takes in the user settings (max transition in this case), other user/design requirements ... and then looks at the library and design data to figure out if they are even reasonable. For now, the changes in this PR are as follows:

- We do check that transition for a specific gate at nominal load is actually going to meet the SDC max transition requirements so the buffer insertion code does not keep trying to fix that error without knowing that it is impossible to fix.
-  There was a genuine bug in the logic where we specify a net of length "X" to be split into lengths "X and 0" (normal expectation is to split it into two non-zero parts. 

The reason both the changes are in the same PR is because both these issues need to be addressed to fix the reported bug. I hope this makes sense.